src/main/java/org/prebid/server/auction/ExchangeService.java

```diff
@@ -156,23 +156,48 @@ public ExchangeService(BidderCatalog bidderCatalog,
 
         final long startTime = clock.millis();
 
-        return extractBidderRequests(bidRequest, uidsCookie, aliases, timeout)
+        // sanity check: discard imps without extension
+        final List<Imp> imps = imps(bidRequest);
+
+        final Map<String, List<String>> deprecatedBiddersErrors = new HashMap<>();
+
+        // identify valid biddersAndErrors and aliases out of imps
+        final List<String> bidders = biddersAndErrors(imps, deprecatedBiddersErrors, aliases);
+
+        return extractBidderRequests(bidRequest, imps, bidders, uidsCookie, aliases, timeout)
                 .map(bidderRequests ->
                         updateRequestMetric(bidderRequests, uidsCookie, aliases, publisherId, metricsContext))
                 .compose(bidderRequests -> CompositeFuture.join(bidderRequests.stream()
                         .map(bidderRequest -> requestBids(bidderRequest, startTime,
                                 auctionTimeout(timeout, cacheInfo.doCaching), aliases, bidAdjustments(requestExt),
                                 currencyRates(targeting)))
                         .collect(Collectors.toList())))
-                // send all the requests to the bidders and gathers results
+                // send all the requests to the biddersAndErrors and gathers results
                 .map(CompositeFuture::<BidderResponse>list)
                 // produce response from bidder results
                 .map(bidderResponses -> updateMetricsFromResponses(bidderResponses, publisherId))
                 .compose(result ->
-                        toBidResponse(result, bidRequest, keywordsCreator, cacheInfo, timeout))
+                        toBidResponse(result, bidRequest, keywordsCreator, cacheInfo, timeout, deprecatedBiddersErrors))
                 .compose(bidResponse -> bidResponsePostProcessor.postProcess(bidRequest, uidsCookie, bidResponse));
     }
 
+    private List<Imp> imps(BidRequest bidRequest) {
+        return bidRequest.getImp().stream()
+                .filter(imp -> imp.getExt() != null)
+                .collect(Collectors.toList());
+    }
+
+    private List<String> biddersAndErrors(List<Imp> imps, Map<String, List<String>> deprecatedBiddersErrors,
+                                          Map<String, String> aliases) {
+        return imps.stream()
+                .flatMap(imp -> asStream(imp.getExt().fieldNames())
+                        .filter(bidder -> !Objects.equals(bidder, PREBID_EXT))
+                        .peek(s -> processDeprecatedBidder(s, deprecatedBiddersErrors))
```

Comment:
I think this code will produce multiple message about the same deprecated bidder if it is present in two or more imps.

In general I would choose another approach to implement this. First of all `holdAuction` and `extractBidderRequests ` method could be left unchanged at all and following line could be added right to `toExtBidResponse`:
```
final Map<...> deprecatedBiddersErrors = bidRequest.getImp().stream()
                .filter(imp -> imp.getExt() != null)
                .flatMap(imp -> asStream(imp.getExt.fieldNames())))
                .distinct()
                .filter(bidder -> bidderCatalog.isDeprecatedName(bidder))
                .collect(Collectors.toMap(Function.identity(), bidder -> Collections.singletonList(bidderCatalog.errorForDeprecatedName(bidder))));
```
As you can see we'll get the same errors map without pushing it through a bunch of methods. This leads to slight duplication but that's ok in this case I think because thereby we've gained cleaner and more focused methods signatures and call chains and that's better for maintenance. Of course if it makes sense common code could be extracted to carefully named methods according to your flavor.